### ISSUE TITLE
Type of Promise not unwrapped with useSelector in suspence mode

### ISSUE THREAD

**Issue by lishine** [LegendApp/legend-state#283]
The example from documentation:

```
import { ObservableObject } from '@legendapp/state';
import { useObservable, useSelector } from '@legendapp/state/react';
import { Suspense } from 'react';

function Test({ state$ }: { state$: ObservableObject<Promise<string>> }) {
  const value = useSelector(state$, { suspense: true });
  return <div>{value}</div>;
}

export default function App() {
  const state$ = useObservable(
    new Promise<string>((resolve) => {
      setTimeout(() => {
        resolve('hello');
      }, 1000);
    })
  );
  return (
    <div>
      <div>Suspense test</div>
      <Suspense fallback={<div>Loading...</div>}>
        <Test state$={state$} />
      </Suspense>
    </div>
  );
}
```

https://stackblitz.com/edit/vitejs-vite-ehsrz6?file=src%2FApp.tsx

Type of `value` should be `string`. Currently it is `Promise<string>`

**Comment by lishine** [comment]
Regarding suspence interface. It would be nice to have something like:
.getSuspended()

**Comment by lishine** [comment]
Turns out the type is different if using `observable` instead of `useObservable`:
`string & WithState`

`observable` and `useObservable` are also different  in that only `useObservable` accepts a function

**Comment by lishine** [comment]
Currently useComputed/computed works differently with promise:
- It has does not return state property
- It does unwrap the type
- It is not reactive to promise fullfillment
- Not sure if it works with suspence

Expect computed to work with promises same as observable of promise:
- Return also the state/_state
- reactive to fullfillment
- unwrap the type
- work with suspence

**Comment by lishine** [comment]
Another problem:
Seems like suspense using `useSelector(observer, { suspense: true }` does not work within a component which wrapped in `observer`

**Comment by lishine** [comment]
I am working on a PR to use promise in `computed`.

**Comment by jmeistrich** [comment]
Before you do that can you try it in the @next version (computed-types branch) which will soon be version 3.0? It changes and improves Promise behavior significantly.

Sorry I haven’t replied yet - things have been in flux with finishing that and I haven’t had a chance to check it with the suspense behavior yet.

**Comment by lishine** [comment]
Reason for suspese not working within `observer` is this shortcircuit:
<img width="862" alt="image" src="https://github.com/LegendApp/legend-state/assets/6741645/458dcf93-908d-4d8f-99c0-d8f386d02f11">

**Comment by lishine** [comment]
3.0 does more changes.
I need to finish fully intergrating legend into our state.
For that enough is to fix current version.
What is left for me is to fix the above suspense issue.

**Comment by lishine** [comment]
Looks like this is the only change for useSelector
<img width="1687" alt="image" src="https://github.com/LegendApp/legend-state/assets/6741645/4d2b4c45-2e43-4c7e-bfdf-3585913a2e45">

**Comment by jmeistrich** [comment]
Ah good catch! That seems like a good fix. Would you like to PR that?

**Comment by lishine** [comment]
> Ah good catch! That seems like a good fix. Would you like to PR that?

Yes

**Comment by lishine** [comment]
https://github.com/LegendApp/legend-state/pull/291
@jmeistrich

**Comment by lishine** [comment]
I did this in order to easier run from the repo directly
![image](https://github.com/LegendApp/legend-state/assets/6741645/0aa264a6-e34e-479a-a4fe-d4b87916fae2)

Better to refactor the imports of `@legendapp/state' in repo into relative imports , I think.
And to build into dist folder.

**Comment by jmeistrich** [comment]
Hmm, I don’t think we want to make such a large change in v2 when we’re just about to release v3. We have a strict rule (especially in v3) that we shouldn’t have have any Promises in observable. v2 was certainly messy and inconsistent with Promise handling, but v3 extracts them out and replaces them with their value, and there’s a `syncState` that can access its status. I think v3 should solve what you’re trying to do very well, and I’m curious how upgrading to it would go?

See this section of the preview docs: https://preview.legend.pages.dev/open-source/state/v3/usage/observable/#async-observables

I would like to get in the suspense fix with credit to you though :). Would you like to submit a PR for just that and we can merge that in easily?

**Comment by lishine** [comment]
Ok, I will look more into v3. Which branch is it?
I tried to make a fix only to `useSelector`,  but it crashes - without fix to computed, but I will try to make it work.

**Comment by lishine** [comment]
I like the `syncState` approach, no need to continue adding properties like `state`.
Just need to make sure that promise works with computed and promised computed works with suspense

**Comment by lishine** [comment]
How do you setup dev env for developing the lib? 
I imported it as a TS into vite app but had to add an alias to babel.

**Comment by lishine** [comment]
@jmeistrich 
PR have been adjusted

**Comment by jmeistrich** [comment]
@lishine It’s the `computed-types` branch but I’m about to merge it into main today 😀. And I merged the adjusted PR, thanks!

For a dev env I currently build it and copy the dist folder into the project’s node_modules. My workflow is mainly: when I find a bug in apps I figure out how to reproduce it, write a test in legend-state that reproduces it and fix it so the test passes, then build and copy to the app to make sure it worked. That’s probably not the most effective way to do it… If you have any suggestions for how to do that better I would appreciate it!

**Comment by jmeistrich** [comment]
That’s merged in and released in the latest 3.0.0-alpha.1. So since that and promise behavior should be working much better now, I’m going to close this issue. But please let me know if it’s still not working correctly and we can re-open this and solve it :).
